contracts/competition/Competition.sol

```diff
@@ -0,0 +1,138 @@
+pragma solidity ^0.6.0;
+
+import "../referrals/interfaces/IReferralStorage.sol";
+
+contract Competition
+{
+    uint public start;
+    uint public end;
+    uint public registrationStart;
+    uint public registrationEnd;
+    IReferralStorage private referralStorage;
+
+    struct Team {
+        address leader;
+        string name;
+        bytes32 referral;
+        address[] members;
+        address[] joinRequests;
+    }
+
+    address[] private leaders;
+    mapping(address => Team) private teams;
+    mapping(string => bool) private teamNames;
+    mapping(address => address) private membersToTeam;
+    mapping(address => mapping(address => bool)) private requests;
+
+    modifier registrationIsOpen()
+    {
+        require(block.timestamp >= registrationStart, "Registration is not opened yet.");
+        require(block.timestamp < registrationEnd, "Registration is closed.");
+        _;
+    }
+
+    modifier isNotLeader()
+    {
+        require(teams[msg.sender].leader == address(0), "Team leaders are not allowed.");
+        _;
+    }
+
+    modifier isNotMember()
+    {
+        require(membersToTeam[msg.sender] == address(0), "Team members are not allowed.");
+        _;
+    }
+
+    constructor (
+        uint _start,
+        uint _end,
+        uint _registrationStart,
+        uint _registrationEnd,
+        IReferralStorage _referralStorage
+    ) public
+    {
+        start = _start;
+        end = _end;
+        registrationStart = _registrationStart;
+        registrationEnd = _registrationEnd;
+        referralStorage = _referralStorage;
+    }
+
+    function registerTeam (string calldata name, bytes32 referral) external registrationIsOpen isNotLeader
+    {
+        require(referralStorage.codeOwners(referral) != address(0), "Referral code does not exist.");
+        require(teamNames[name] == false, "Team name already registered.");
+
+        Team storage team;
+
+        team.leader = msg.sender;
+        team.name = name;
+        team.referral = referral;
+        team.members.push(msg.sender);
+
+        teams[msg.sender] = team;
+        leaders.push(msg.sender);
+        teamNames[name] = true;
+    }
+
+    function createJoinRequest (address leaderAddress) external registrationIsOpen isNotLeader isNotMember
```

Comment:
i think `isNotMember` is sufficient since each leader is member

Reply:
Make sense, will remove it